### PR TITLE
storage: bugfixes

### PR DESCRIPTION
**Pull request by brauner**

### CODE REVIEW

File: lxd/patches.go [hunk 1]
```diff
@@ -37,7 +38,7 @@ var patches = []patch{
 	{name: "storage_api", run: patchStorageApi},
 	{name: "storage_api_v1", run: patchStorageApiV1},
 	{name: "storage_api_dir_cleanup", run: patchStorageApiDirCleanup},
-	{name: "storage_api_lvm_keys", run: patchStorageApiLvmKeys},
+	{name: "storage_api_zfs_lvm_keys", run: patchStorageApiZfsLvmKeys},
```
Comment (stgraber) [inline lxd/patches.go]: Please make this a separate patch and keep storage_api_lvm_keys as it is. Otherwise the few folks who use git master will end up with a patch in their database that doesn't have any matching code (not going to break anything, but annoying for debugging).

File: lxd/storage_lvm.go [hunk 2]
```diff
@@ -238,23 +238,27 @@ func (s *storageLvm) StoragePoolInit(config map[string]interface{}) (storage, er
...
+				// Internal error.
+				return s, err
+			} else if !ok {
+				// Volume group does not exist.
+				return s, fmt.Errorf("The requested volume group \"%s\" does not exist.", source)
+			}
 		}
-		s.log.Debug(fmt.Sprintf("Source property of this lvm storage pool empty: Detected that volume group \"lvm.vg_name=%s\" already exists.", s.vgName))
-	} else if filepath.IsAbs(source) && !shared.IsBlockdevPath(source) {
-		return s, fmt.Errorf("Loop backed lvm storage volumes are currently not supported.")
```
Comment (stgraber) [inline lxd/storage_lvm.go]: Do we have an issue open for loop backed LVM?

File: lxd/storage_pools_config.go [hunk 3]
```diff
@@ -50,8 +50,6 @@ func storagePoolValidateConfig(name string, driver string, config map[string]str
 	if config["source"] == "" {
 		if driver == "dir" {
 			config["source"] = filepath.Join(shared.VarPath("storage-pools"), name)
-		} else if driver != "lvm" {
-			config["source"] = filepath.Join(shared.VarPath("disks"), name)
 		}
 	}
```
Comment (stgraber) [inline lxd/storage_pools_config.go]: storagePoolValidateConfig really shouldn't ever modify the config, this should go into storagePoolFillDefault instead.

Comment (stgraber) [inline lxd/storage_pools_config.go]: My branch makes sure that storagePoolValidateConfig doesn't call storagePoolFillDefault as it's otherwise breaking pool configuration when LXD decides to fill in read-only properties

File: lxd/storage_zfs.go [hunk 4]
```diff
@@ -1341,7 +1312,13 @@ func (s *storageZfs) zfsPoolCreate() error {
 		vdev = filepath.Join(shared.VarPath("disks"), s.pool.Name)
 	}
 
+	zpoolName := s.pool.Name
 	if !filepath.IsAbs(vdev) {
+		s.dataset = vdev
+		if s.pool.Config["zfs.pool_name"] != "" {
+			return fmt.Errorf("Invalid request.")
+		}
+		s.pool.Config["zfs.pool_name"] = vdev
```
Comment (stgraber) [inline lxd/storage_zfs.go]: Does that actually hit the database?

File: lxd/storage_zfs.go [hunk 5]
```diff
@@ -1399,12 +1380,8 @@ func (s *storageZfs) zfsPoolCreate() error {
 			// UUID in a multi-device pool for all devices.). The
 			// safest way is to just store the name of the zfs pool
 			// we create.
-			s.pool.Config["source"] = s.pool.Name
-		}
-
-		zpoolName := s.pool.Config["zfs.pool_name"]
-		if zpoolName == "" {
-			zpoolName = s.pool.Name
+			s.pool.Config["source"] = zpoolName
```
Comment (stgraber) [inline lxd/storage_zfs.go]: Same question as above, does that get saved in the database?

File: lxd/storage_lvm.go [hunk 6]
```diff
@@ -324,65 +326,66 @@ func (s *storageLvm) StoragePoolCreate() error {
...
+			if err == nil && !ok {
+				// Create a volume group on the physical volume.
+				output, err := exec.Command("vgcreate", poolName, source).CombinedOutput()
+				if err != nil {
+					return fmt.Errorf("Failed to create the volume group for the lvm storage pool: %s.", output)
 				}
-			}()
-		}
+			}
+			s.pool.Config["lvm.vg_name"] = poolName
```
Comment (stgraber) [inline lxd/storage_lvm.go]: Same question as zfs, does that end up in the database?

File: lxd/storage_lvm.go [hunk 7]
```diff
@@ -324,65 +326,66 @@ func (s *storageLvm) StoragePoolCreate() error {
...
+
+			// lxc storage create pool1 lvm source=/dev/sdX lvm.vg_name=random-string
+			poolName = s.pool.Config["lvm.vg_name"]
+			if poolName == "" {
+				// lxc storage create pool1 lvm source=/dev/sdX
+				poolName = s.pool.Name
+			}
+
+			// Set source to volume group name.
+			s.pool.Config["source"] = poolName
```
Comment (stgraber) [inline lxd/storage_lvm.go]: Same question as zfs, does that end up in the database?

File: lxd/storage_btrfs.go [hunk 8]
```diff
@@ -100,7 +100,8 @@ func (s *storageBtrfs) StoragePoolCheck() error {
 func (s *storageBtrfs) StoragePoolCreate() error {
 	source := s.pool.Config["source"]
 	if source == "" {
-		return fmt.Errorf("No \"source\" property found for the storage pool.")
+		source = filepath.Join(shared.VarPath("disks"), s.pool.Name)
+		s.pool.Config["source"] = source
```
Comment (stgraber) [inline lxd/storage_btrfs.go]: Same question as zfs, does that end up in the database?

File: lxd/storage_zfs.go (hunk 4 above)
Comment (brauner) [inline lxd/storage_zfs.go]: Yes, after pool creation the config is diffed again

File: lxd/patches.go (hunk 1 above)
Comment (brauner) [inline lxd/patches.go]: Right

File: lxd/storage_lvm.go (hunk 2 above)
Comment (brauner) [inline lxd/storage_lvm.go]: No, and we decided not to on purpose until we get explicit user requests a while back.

File: lxd/storage_pools_config.go (hunk 3 above)
Comment (brauner) [inline lxd/storage_pools_config.go]: Will rebase on yours once I've reviewed and merged.

File: lxd/storage_zfs.go (hunk 5 above)
Comment (brauner) [inline lxd/storage_zfs.go]: Yes, after pool creation the config is diffed again

File: lxd/storage_lvm.go (hunk 6 above)
Comment (brauner) [inline lxd/storage_lvm.go]: Yes, after pool creation the config is diffed again

File: lxd/storage_lvm.go (hunk 7 above)
Comment (brauner) [inline lxd/storage_lvm.go]: Yes, after pool creation the config is diffed again

File: lxd/storage_btrfs.go (hunk 8 above)
Comment (brauner) [inline lxd/storage_btrfs.go]: Yes, after pool creation the config is diffed again

File: lxd/patches.go (hunk 1 above)
Comment (brauner) [inline lxd/patches.go]: Fixed.

**Comment by brauner** [comment]
Closes #2928.

File: lxd/storage_lvm.go (hunk 2 above)
Comment (stgraber) [inline lxd/storage_lvm.go]: @brauner yeah, not saying we need to rush it, but it sounds pretty easy to do considering that the code is already there for btrfs and it'd make our testsuite quite a bit cleaner too.

Comment (brauner) [inline lxd/storage_lvm.go]: There originally was a legitimate reason to punt on this: I think it had something to do with lvm device caching which doesn't work with loop devices or something. IRC logs should reveal the real reason.

File: lxd/patches.go [hunk 9]
```diff
@@ -345,7 +347,15 @@ func upgradeFromStorageTypeBtrfs(name string, d *Daemon, defaultPoolName string,
...
-				return err
+				err = btrfsSnapshot(oldContainerMntPoint, newContainerMntPoint, false)
+				if err != nil {
+					return err
+				}
+				err = btrfsSubVolumesDelete(oldContainerMntPoint)
+				if err != nil {
+					btrfsSubVolumesDelete(newContainerMntPoint)
+					return err
+				}
```
Comment (stgraber) [inline lxd/patches.go]: Why is that code needed? when would Rename fail but a subvolume snapshot succeed?

**Comment by brauner** [comment]
Closes #2929.

File: lxd/patches.go (hunk 9 above)
Comment (brauner) [inline lxd/patches.go]: btrfs subvolume's should be able to handle cross-device links in case the btrfs pool uses multiple devices.

File: lxd/patches.go [hunk 10]
```diff
@@ -476,7 +486,15 @@ func upgradeFromStorageTypeBtrfs(name string, d *Daemon, defaultPoolName string,
...
-				return err
+				err = btrfsSnapshot(oldImageMntPoint, newImageMntPoint, false)
+				if err != nil {
+					return err
+				}
+				err = btrfsSubVolumesDelete(oldImageMntPoint)
+				if err != nil {
+					btrfsSubVolumesDelete(newImageMntPoint)
+					return err
+				}
```
Comment (stgraber) [inline lxd/patches.go]: Same here. btrfs snapshot can't be cross device so I'm not sure when rename would fail but snapshot wouldn't.

Comment (brauner) [inline lxd/patches.go]: Hm, ah snapshots... So we could try to do a btrfssubvolumecreate and rsync over
but it might not be worth the trouble

File: lxd/patches.go (hunk 9 above)
Comment (brauner) [inline lxd/patches.go]: Hm, ah snapshots... So we could try to do a btrfssubvolumecreate and rsync over
but it might not be worth the trouble

File: lxd/patches.go [hunk 11]
```diff
@@ -986,7 +1021,15 @@ func upgradeFromStorageTypeZfs(name string, d *Daemon, defaultPoolName string, d
 			poolConfig["source"] = shared.VarPath("disks", defaultPoolName+".img")
 			err := os.Rename(oldLoopFilePath, poolConfig["source"])
 			if err != nil {
-				return err
+				output, err := storageRsyncCopy(oldLoopFilePath, poolConfig["source"])
```
Comment (stgraber) [inline lxd/patches.go]: should be shared.FileMove

Comment (brauner) [inline lxd/patches.go]: shared.FileMove will only move files, not copy directories including contents, no?

Comment (brauner) [inline lxd/patches.go]: and only a single one at that

Comment (stgraber) [inline lxd/patches.go]: You're moving a single file here, that's why I'm saying you can use FileMove()

File: lxd/container.go [hunk 12]
```diff
@@ -748,22 +748,24 @@ func containerCreateInternal(d *Daemon, args containerArgs) (container, error) {
...
-	for _, m := range c.ExpandedDevices() {
-		if m["type"] != "disk" || m["path"] != "/" || m["size"] == "" {
-			continue
-		}
+	localDevices := c.LocalDevices()
+	_, rootDiskDevice := containerGetRootDiskDevice(localDevices)
+	if rootDiskDevice["pool"] == "" {
+		expandedDevices := c.ExpandedDevices()
+		_, rootDiskDevice = containerGetRootDiskDevice(expandedDevices)
+	}
```
Comment (stgraber) [inline lxd/container.go]: Hmm, what?

_, rootDiskDevice := containerGetRootDiskDevice(c.ExpandedDevices())

Should do exactly that right?

Comment (brauner) [inline lxd/container.go]: Yeah, I just don't like nesting function calls that much...

File: lxd/storage_btrfs.go [hunk 13]
```diff
@@ -1251,7 +1253,11 @@ func (s *storageBtrfs) btrfsPoolVolumeCreate(subvol string) error {
 	return nil
 }
 
-func (s *storageBtrfs) btrfsPoolVolumeQGroup(subvol string) (string, error) {
+func (s *storageBtrfs) btrfsPoolVolumeCreate(subvol string) error {
+	return btrfsSubVolumeCreate(subvol)
+}
```
Comment (stgraber) [inline lxd/storage_btrfs.go]: Can't we do away with those wrapper functions and just call btrfsSubvolumeCreate directly?

Comment (stgraber) [inline lxd/storage_btrfs.go]: Same apply to all the others

File: lxd/storage_btrfs.go [hunk 14]
```diff
@@ -1319,82 +1329,70 @@ func (s *storageBtrfs) btrfsPoolVolumeQGroupUsage(subvol string) (int64, error)
...
-	).CombinedOutput()
+	).Run()
 
-	if err != nil {
-		s.log.Warn(
-			"subvolume delete failed",
-			log.Ctx{"subvol": subvol, "output": string(output)},
-		)
-	}
 	return nil
```
Comment (stgraber) [inline lxd/storage_btrfs.go]: Why are we ignoring all errors in there now?

Comment (stgraber) [inline lxd/storage_btrfs.go]: Not really sure why we were just treating them as warnings in the first place too, I'd expect this function to fail if something actually fails...

Comment (brauner) [inline lxd/storage_btrfs.go]: We did before apart from logging. But I can add that back.

File: lxd/storage_lvm.go [hunk 15]
```diff
@@ -238,23 +238,34 @@ func (s *storageLvm) StoragePoolInit(config map[string]interface{}) (storage, er
 	}
 
 	source := s.pool.Config["source"]
-	s.vgName = s.pool.Config["lvm.vg_name"]
 	s.thinPoolName = s.pool.Config["lvm.thinpool_name"]
+	if s.thinPoolName == "" {
+		// Should actually always be set. But let's be sure.
+		s.thinPoolName = "LXDThinpool"
+	}
+	s.thinPoolName = s.pool.Config["lvm.thinpool_name"]
```
Comment (stgraber) [inline lxd/storage_lvm.go]: That makes not sense :)

You're setting s.thinPoolName from the Config, then check if it's empty, if it is, you set it with the default value, which you then overwrite with the same empty value from the Config

File: lxd/storage_pools_config.go (hunk 3 above)
Comment (stgraber) [inline lxd/storage_pools_config.go]: This code should still be moved away from ValidateConfig

File: lxd/storage_lvm.go (hunk 15 above)
Comment (brauner) [inline lxd/storage_lvm.go]: #ThingsYouShouldntDoInTheMiddleOfTheNight

File: lxd/storage_zfs.go [hunk 16]
```diff
@@ -342,10 +309,6 @@ func (s *storageZfs) StoragePoolVolumeUpdate(changedConfig []string) error {
 		return fmt.Errorf("The \"size\" property cannot be changed.")
 	}
 
-	if shared.StringInSlice("zfs.use_refquota", changedConfig) {
-		return fmt.Errorf("The \"zfs.use_refquota\" property cannot be changed.")
-	}
-
 	if shared.StringInSlice("zfs.remove_snapshots", changedConfig) {
 		return fmt.Errorf("The \"zfs.remove_snapshots\" property cannot be changed.")
 	}
```
Comment (stgraber) [inline lxd/storage_zfs.go]: This one should also be user modifiable.

File: lxd/storage_zfs.go [hunk 17]
```diff
@@ -314,10 +285,6 @@ func (s *storageZfs) StoragePoolUpdate(changedConfig []string) error {
 		return fmt.Errorf("The \"lvm.vg_name\" property cannot be changed.")
 	}
 
-	if shared.StringInSlice("volume.zfs.use_refquota", changedConfig) {
-		return fmt.Errorf("The \"volume.zfs.use_refquota\" property cannot be changed.")
-	}
-
 	if shared.StringInSlice("volume.zfs.remove_snapshots", changedConfig) {
 		return fmt.Errorf("The \"volume.zfs.remove_snapshots\" property cannot be changed.")
 	}
```
Comment (stgraber) [inline lxd/storage_zfs.go]: This one should also be user modifiable.

File: lxd/storage_btrfs.go [hunk 18]
```diff
@@ -98,68 +98,69 @@ func (s *storageBtrfs) StoragePoolCheck() error {
...
+					err := s.btrfsPoolVolumeCreate(source)
+					if err != nil {
+						return err
+					}
+				} else {
+					return fmt.Errorf("Custom loop file locations are not supported.")
+				}
 			}
+		} else {
+			return fmt.Errorf("Invalid request.")
```
Comment (stgraber) [inline lxd/storage_btrfs.go]: Maybe be a bit more specific

File: lxd/storage_btrfs.go (hunk 14 above)
Comment (stgraber) [inline lxd/storage_btrfs.go]: Yeah, I'm assuming we're not ignoring errors in the other storage backends, so we shouldn't here either.

File: lxd/storage_btrfs.go (hunk 18 above)
Comment (brauner) [inline lxd/storage_btrfs.go]: Well, that could actually be a lot. We can report back the values the user gave...

File: lxd/container.go (hunk 12 above)
Comment (stgraber) [inline lxd/container.go]: ah, I think you misunderstood. I didn't meant to just change that one line. I meant you can replace the whole hunk with just that line since c.ExpandedDevices() will always return the local entry if there's one.

Comment (stgraber) [inline lxd/container.go]: So:
```
	localDevices := c.LocalDevices()
	_, rootDiskDevice := containerGetRootDiskDevice(localDevices)
	if rootDiskDevice["pool"] == "" {
		expandedDevices := c.ExpandedDevices()
		_, rootDiskDevice = containerGetRootDiskDevice(expandedDevices)
	}
```

becomes just:

```
    _, rootDiskDevice := containerGetRootDiskDevice(c.ExpandedDevices())
```

File: lxd/storage_btrfs.go (hunk 18 above)
Comment (stgraber) [inline lxd/storage_btrfs.go]: "Invalid source path" would be an improvement

File: lxd/container.go (hunk 12 above)
Comment (brauner) [inline lxd/container.go]: No, because the local devices could have a root disk entry that has size property set whereas the profile device hasn't.

Comment (brauner) [inline lxd/container.go]: If I'm thinking straight.

File: lxd/storage_btrfs.go (hunk 13 above)
Comment (brauner) [inline lxd/storage_btrfs.go]: removed.

File: lxd/storage_btrfs.go (hunk 18 above)
Comment (brauner) [inline lxd/storage_btrfs.go]: Pah :)

File: lxd/container.go (hunk 12 above)
Comment (stgraber) [inline lxd/container.go]: So? If the device exists in LocalDevices, that's what you'll end up with in ExpandedDevices.

Comment (stgraber) [inline lxd/container.go]: ExpandedDevices applies each profile in sequence, overriding entire device entries (not device properties) when another one of the same name is found. It then applies LocalDevices on top of that, overriding anything that came from a profile if the container has it in its own devices.

That's how the whole inheritance works for Devices and Config throughout LXD.

Comment (brauner) [inline lxd/container.go]: I thought last time we established that if you had a device:

root:
  type: disk
  path: /
  size: 100GB

In the containers config and a device called:

defintely-another-name:
  type: disk
  path: /

in the profiles then you end up with both devices:
- root
- definitely-another-name
in the expanded devices...

Comment (stgraber) [inline lxd/container.go]: Right, you can indeed end up in that case because we don't want to prevent profile config changes because of container local configuration.

But such a container then becomes invalid and won't start as it'll fail the rootfs check then (more than one path=/).

Comment (brauner) [inline lxd/container.go]: I won't now, because the check for the same path twice is currently not performed for that exact same reason on the expanded devices.

Comment (stgraber) [inline lxd/container.go]: Fine, we'll do that separately, this branch is giving me a headache as it is. But we'll need this sorted before 2.9.3.

File: lxd/storage_pools_config.go (hunk 3 above)
Comment (brauner) [inline lxd/storage_pools_config.go]: removed. (i though you did all that in your branch.)

File: lxd/container.go (hunk 12 above)
Comment (brauner) [inline lxd/container.go]: i'll send another patch after this one

File: lxd/storage_btrfs.go (hunk 14 above)
Comment (brauner) [inline lxd/storage_btrfs.go]: nope, it was only the qgroup stuff removal stuff

File: lxd/storage_zfs.go (hunk 16 above)
Comment (brauner) [inline lxd/storage_zfs.go]: modifiable

File: lxd/storage_zfs.go (hunk 17 above)
Comment (brauner) [inline lxd/storage_zfs.go]: modifiable

File: lxd/storage_btrfs.go (hunk 14 above)
Comment (stgraber) [inline lxd/storage_btrfs.go]: I'm tempted to say that even the qgroup stuff should be fatal if it fails. We only run that code if we detected a qgroup exists, so if we fail to remove it, there's something wrong.

File: lxd/storage_btrfs.go (hunk 18 above)
Comment (brauner) [inline lxd/storage_btrfs.go]: changed to something more useful (sort of).
